### ISSUE TITLE
✨ consider writing a shuffle service 

### ISSUE THREAD

**Issue by joocer** [mabel-dev/opteryx#1752]
A remote shuffle service can indeed be implemented using REST APIs, but it generally involves a combination of architectural components to handle the complexity and requirements of distributed data processing. Here's a high-level overview of what a remote shuffle service might look like:

### Components of a Remote Shuffle Service

1. **Shuffle Servers**:
   - Dedicated servers or processes responsible for managing the shuffle data. These servers handle the storage, retrieval, and transfer of shuffle data between executors.

2. **Data Storage**:
   - A distributed storage system (e.g., HDFS, S3, or a custom distributed file system) where the shuffle data is stored. This storage system should be optimized for high I/O operations and fault tolerance.

3. **Metadata Management**:
   - A metadata service that keeps track of where the shuffle data is stored, ensuring that it can be efficiently located and retrieved. This service could use a distributed key-value store (e.g., ZooKeeper, etcd).

### APIs and Protocols

1. **Data Write API**:
   - Executors send their shuffle data to the shuffle servers using a write API. This API allows executors to write shuffle data blocks, specifying metadata such as shuffle ID, partition ID, and other relevant information.
   - **Example**:
     ```http
     POST /shuffle/write
     {
       "shuffleId": "123",
       "partitionId": "456",
       "data": "base64encodeddata..."
     }
     ```

2. **Data Read API**:
   - Executors or other components request shuffle data from the shuffle servers using a read API. This API allows retrieving specific partitions of shuffle data needed for the next stage of processing.
   - **Example**:
     ```http
     GET /shuffle/read?shuffleId=123&partitionId=456
     ```

3. **Metadata Management API**:
   - APIs to manage and query metadata about the shuffle data. This includes registering shuffle data locations, querying shuffle status, and managing the lifecycle of shuffle data.
   - **Example**:
     ```http
     GET /shuffle/metadata?shuffleId=123
     ```

4. **Health and Monitoring APIs**:
   - APIs for health checks, status monitoring, and performance metrics of the shuffle servers to ensure the service is running smoothly.
   - **Example**:
     ```http
     GET /shuffle/health
     ```

### Workflow

1. **Shuffle Write**:
   - Executors process tasks and generate intermediate shuffle data. This data is sent to the shuffle servers using the Data Write API. The shuffle servers store this data in the distributed storage and update the metadata service.

2. **Shuffle Read**:
   - During the next stage of processing, executors request the necessary shuffle data partitions using the Data Read API. The shuffle servers retrieve the data from storage and send it to the requesting executors.

3. **Metadata Management**:
   - The metadata service tracks all shuffle data locations and statuses, ensuring efficient data retrieval and management. Executors query this service to locate shuffle data.

4. **Monitoring and Maintenance**:
   - Health and monitoring APIs provide insights into the performance and health of the shuffle servers, allowing for proactive maintenance and scaling.

### Implementation Considerations

- **Performance**: The APIs and underlying storage systems must be optimized for high throughput and low latency to handle the demands of shuffle operations.
- **Scalability**: The system should be able to scale horizontally to handle increasing amounts of shuffle data.
- **Fault Tolerance**: Data replication and robust error handling mechanisms are essential to ensure data reliability and service availability.
- **Security**: Authentication, authorization, and encryption mechanisms should be in place to protect data and ensure secure communications between components.

In summary, while REST APIs can be a part of a remote shuffle service, the overall architecture involves several components working together to efficiently manage shuffle data in a distributed processing environment.
